batch/batch/front_end/front_end.py

```diff
@@ -1418,46 +1449,53 @@ async def delete(tx):
     return web.HTTPFound(deploy_config.external_url('batch', '/billing_projects'))
 
 
-@routes.post('/billing_projects/{billing_project}/users/add')
-@prom_async_time(REQUEST_TIME_POST_BILLING_PROJECT_ADD_USER_UI)
-@check_csrf_token
-@web_authenticated_developers_only(redirect=False)
-async def post_billing_projects_add_user(request, userdata):  # pylint: disable=unused-argument
-    db = request.app['db']
-    post = await request.post()
-    user = post['user']
-    billing_project = request.match_info['billing_project']
-
-    session = await aiohttp_session.get_session(request)
-
+async def _add_user_to_billing_project(db, billing_project, user):
     @transaction(db)
     async def insert(tx):
         row = await tx.execute_and_fetchone(
             '''
-SELECT billing_projects.name as billing_project, user
-FROM billing_projects
+SELECT billing_projects.name as billing_project,
+billing_projects.`status` as `status`,
+user FROM billing_projects
 LEFT JOIN (SELECT * FROM billing_project_users
-    WHERE billing_project = %s AND user = %s FOR UPDATE) AS t
-  ON billing_projects.name = t.billing_project
-WHERE billing_projects.name = %s;
-''',
+WHERE billing_project = %s AND user = %s FOR UPDATE) AS t
+ON billing_projects.name = t.billing_project
+WHERE billing_projects.name = %s AND billing_projects.`status` != 'deleted';
+        ''',
```

Comment:
I think I'd actually do this query as:

```sql
SELECT billing_projects.name as billing_project,
       billing_projects.`status` as `status`,
       user
FROM billing_projects
INNER JOIN billing_project_users
ON billing_projects.name = billing_project_users.billing_project
WHERE billing_projects.name = %s
  AND billing_projects.`status` != 'deleted'
  AND billing_projects.user = %s
FOR UPDATE
```

Reply:
Hm… I don't need to lock the `billing_projects` for update, though---seems perfectly valid to e.g. simultaneously add a bunch of users to the same billing project.

Reply:
Can I write this join with slightly more granularity on the locks?

Reply:
I want to preserve the behavior where we pop up the different messages when a user is added through the UI (although I'll do this for the API version)---currently neither are errors, but one will tell the user "X has been added to billing project Y" vs "X is already a member of billing project Y".

Also, as discussed, I'm keeping the subquery structure here since it was already there (but putting a shared lock on the billing projects table), making the update to the batches table query as you suggested below, and we'll revisit this in a separate PR.